### PR TITLE
Name explicit registers in conflict register errors for inline assembly

### PR DESCRIPTION
**Pull request by GeorgeWort**

### CODE REVIEW

File: compiler/rustc_ast_lowering/src/asm.rs [hunk 1]
```diff
@@ -374,11 +374,28 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
...
+                                            | InlineAsmOperand::Out { reg, .. }
+                                            | InlineAsmOperand::InOut { reg, .. }
+                                            | InlineAsmOperand::SplitInOut { reg, .. } => {
+                                                if let InlineAsmRegOrRegClass::Reg(reg_sym) = reg {
+                                                    reg_sym.as_str()
+                                                } else {
+                                                    default
+                                                }
+                                            }
+                                            _ => default,
```
Comment (petrochenkov) [inline compiler/rustc_ast_lowering/src/asm.rs:390]:
```suggestion
                                            _ => unreachable!(),
```
I don't think the `default` is ever necessary here, both branches using `default` are unreachable.

File: compiler/rustc_ast_lowering/src/asm.rs [hunk 2]
```diff
@@ -374,11 +374,28 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
                                         _ => None,
                                     };
 
+                                    let reg_str = |idx, default| -> &str {
+                                        let (op, _): &(InlineAsmOperand, Span) = &asm.operands[idx];
+                                        match op {
```
Comment (petrochenkov) [inline compiler/rustc_ast_lowering/src/asm.rs:379]:
```suggestion
                                        match asm.operands[idx].0 {
```

Comment (petrochenkov) [reply compiler/rustc_ast_lowering/src/asm.rs:379]: Could you also add a comment saying that HIR asm doesn't preserve the original alias spelling, so we go to AST to get it?

**Comment by petrochenkov** [comment]
Some more tests in `tests/ui/asm` also need an update.
@rustbot author

**Comment by petrochenkov** [comment]
r=me with the style nit addressed https://github.com/rust-lang/rust/pull/117912#discussion_r1407692702

File: compiler/rustc_ast_lowering/src/asm.rs [hunk 3]
```diff
@@ -374,11 +370,24 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
                                         _ => None,
                                     };
 
+                                    let reg_str = |idx| -> &str {
+                                        // HIR asm doesn't preserve the original alias string of the explicit register,
+                                        // so we have to retrieve it from AST
+                                        let (op, _): &(InlineAsmOperand, Span) = &asm.operands[idx];
+                                        if let Some(ast::InlineAsmRegOrRegClass::Reg(reg_sym)) =
```
[the remark this answers is not in the document]
Comment (GeorgeWort) [reply compiler/rustc_ast_lowering/src/asm.rs:377]: The `let (op, )` exists as the compiler cannot infer the type of the asm.operands access, and `./x fmt` adds braces onto the `Some` branch turning it into 7 lines anyway, so is it still worth using a match?

**Comment by petrochenkov** [comment]
(CI appears to be down in general.)

**Pull request merged**